File: DSA/GreedyAlgo/Code/ActivitySelection.py
# Activity Selection Problem in Python

class Activity:
    def __init__(self, id, start, end):
        self.id = id
        self.start = start
        self.end = end

def activity_selection(activities):
    # Sort by end time
    activities.sort(key=lambda x: x.end)
    
    print("Selected Activities:")
    print(f"Activity {activities[0].id} ({activities[0].start} - {activities[0].end})")
    
    last_end = activities[0].end
    count = 1
    
    for i in range(1, len(activities)):
        if activities[i].start >= last_end:
            print(f"Activity {activities[i].id} ({activities[i].start} - {activities[i].end})")
            last_end = activities[i].end
            count += 1
    
    print(f"\nTotal Activities Selected: {count}")

# Test
print("=== Activity Selection Problem (Python) ===\n")

activities = [
    Activity(1, 0, 5),
    Activity(2, 1, 2),
    Activity(3, 3, 4),
    Activity(4, 0, 6),
    Activity(5, 5, 7),
    Activity(6, 8, 9)
]

print("Activities (ID, Start, End):")
for a in activities:
    print(f"Activity {a.id}: {a.start} - {a.end}")
print()

activity_selection(activities)

print("\n=== Greedy Strategy ===")
print("1. Sort activities by end time")
print("2. Select first activity")
print("3. For remaining activities:")
print("   - If start >= last end: select it")
print("4. Repeat until all checked\n")

print("=== Complexity ===")
print("Time Complexity:  O(n log n) for sorting")
print("Space Complexity: O(1)")
print("Greedy Choice: Earliest finish time")

File: DSA/GreedyAlgo/Code/ActivitySelection.cs
// Activity Selection Problem in C#

using System;
using System.Collections.Generic;
using System.Linq;

class ActivitySelection {
    class Activity : IComparable<Activity> {
        public int Id { get; set; }
        public int Start { get; set; }
        public int End { get; set; }
        
        public Activity(int id, int start, int end) {
            Id = id;
            Start = start;
            End = end;
        }
        
        public int CompareTo(Activity other) {
            return this.End - other.End;
        }
    }
    
    public static void ActivitySelectionProblem(List<Activity> activities) {
        activities.Sort();
        
        Console.WriteLine("Selected Activities:");
        Console.WriteLine($"Activity {activities[0].Id} ({activities[0].Start} - {activities[0].End})");
        
        int lastEnd = activities[0].End;
        int count = 1;
        
        for (int i = 1; i < activities.Count; i++) {
            if (activities[i].Start >= lastEnd) {
                Console.WriteLine($"Activity {activities[i].Id} ({activities[i].Start} - {activities[i].End})");
                lastEnd = activities[i].End;
                count++;
            }
        }
        
        Console.WriteLine($"\nTotal Activities Selected: {count}");
    }
    
    static void Main() {
        Console.WriteLine("=== Activity Selection Problem (C#) ===\n");
        
        List<Activity> activities = new List<Activity> {
            new Activity(1, 0, 5),
            new Activity(2, 1, 2),
            new Activity(3, 3, 4),
            new Activity(4, 0, 6),
            new Activity(5, 5, 7),
            new Activity(6, 8, 9)
        };
        
        Console.WriteLine("Activities (ID, Start, End):");
        foreach (var a in activities) {
            Console.WriteLine($"Activity {a.Id}: {a.Start} - {a.End}");
        }
        Console.WriteLine();
        
        ActivitySelectionProblem(activities);
        
        Console.WriteLine("\n=== Greedy Strategy ===");
        Console.WriteLine("1. Sort activities by end time");
        Console.WriteLine("2. Select first activity");
        Console.WriteLine("3. For remaining activities:");
        Console.WriteLine("   - If start >= last end: select it");
        Console.WriteLine("4. Repeat until all checked\n");
        
        Console.WriteLine("=== Complexity ===");
        Console.WriteLine("Time Complexity:  O(n log n) for sorting");
        Console.WriteLine("Space Complexity: O(1)");
        Console.WriteLine("Greedy Choice: Earliest finish time");
    }
}
